pkg/schedule/affinity/group.go

```diff
@@ -0,0 +1,380 @@
+//   - newState = groupAvailable: change the state to groupAvailable
+//   - newState = groupDegraded: do nothing and do not refresh degradedExpiredAt
+//   - newState = groupExpired: change the state to groupExpired
+//
+// - When the current state is groupExpired
+//   - newState = groupAvailable: change the state to groupAvailable
+//   - newState = groupDegraded: do nothing and do not refresh degradedExpiredAt
+//   - newState = groupExpired: do nothing
+func (g *runtimeGroupInfo) SetState(newState condition) {
+	// If the expiration time has been reached, change groupDegraded to groupExpired.
+	if g.state == groupDegraded && g.IsExpired() {
+		g.state = groupExpired
+	}
+	// Update state
+	newState = newState.toGroupState()
+	if newState == groupDegraded {
+		// Only set the expiration time when transitioning from groupAvailable to groupDegraded.
+		// Do nothing if the original state is already groupDegraded or groupExpired.
+		if g.state == groupAvailable {
+			g.state = groupDegraded
+			g.degradedExpiredAt = newDegradedExpiredAtFromNow()
+		}
+	} else {
+		g.state = newState
+	}
+}
+
+func newDegradedExpiredAtFromNow() uint64 {
+	return uint64(time.Now().Unix()) + defaultDegradedExpirationSeconds
+}
+
+// IsAffinitySchedulingEnabled indicates whether affinity scheduling is allowed.
+func (g *runtimeGroupInfo) IsAffinitySchedulingEnabled() bool {
+	return g.IsAvailable() && g.LeaderStoreID != 0 && len(g.VoterStoreIDs) != 0
+}
+
+// IsRegularSchedulingEnabled indicates whether balance scheduling is allowed.
+func (g *runtimeGroupInfo) IsRegularSchedulingEnabled() bool {
+	return g.IsExpired() || g.LeaderStoreID == 0 || len(g.VoterStoreIDs) == 0
+}
+
+// AdjustGroup validates the group and sets default values.
+func (m *Manager) AdjustGroup(g *Group) error {
+	if err := ValidateGroupID(g.ID); err != nil {
+		return err
+	}
+	// If no distribution is provided, we will use the default distribution.
+	if g.LeaderStoreID == 0 && len(g.VoterStoreIDs) == 0 {
+		return nil
+	}
+	// Once provided, leader store ID and voter store IDs must be provided together.
+	if g.LeaderStoreID == 0 || len(g.VoterStoreIDs) == 0 {
+		return errs.ErrAffinityGroupContent.FastGenByArgs("leader store ID and voter store IDs must be provided together")
+	}
+
+	voterStoreIDs := slices.Clone(g.VoterStoreIDs)
+	slices.Sort(voterStoreIDs)
+	if slice.HasDupInSorted(voterStoreIDs) {
+		return errs.ErrAffinityGroupContent.FastGenByArgs("duplicate voter store ID")
+	}
+	if !slices.Contains(voterStoreIDs, g.LeaderStoreID) {
+		return errs.ErrAffinityGroupContent.FastGenByArgs("leader must be in voter stores")
+	}
+	for _, storeID := range voterStoreIDs {
+		if m.storeSetInformer.GetStore(storeID) == nil {
+			return errs.ErrAffinityGroupContent.FastGenByArgs("store does not exist")
+		}
+	}
+
+	return nil
+}
+
+// GroupKeyRanges represents key ranges with group id.
+type GroupKeyRanges struct {
+	KeyRanges []keyutil.KeyRange
+	GroupID   string
```

Comment:
How about adding more to the format of this groupID?

Reply:
Or we add it in pd-ctl about group format, I have added it in pd-ctl. We don't limit the format in server, it should be a convention for pd-ctl and SQL.

Reply:
The region label also does not conform to the server's specified format.